ext/opentelemetry-exporter-cloud-trace/src/opentelemetry/exporter/cloud_trace/__init__.py

```diff
@@ -0,0 +1,284 @@
+
+            cloud_trace_spans.append(
+                {
+                    "name": span_name,
+                    "span_id": span_id,
+                    "display_name": _get_truncatable_str_object(
+                        span.name, 128
+                    ),
+                    "start_time": start_time,
+                    "end_time": end_time,
+                    "parent_span_id": parent_id,
+                    "attributes": attributes,
+                    "links": _extract_links(span.links),
+                    "status": _extract_status(span.status),
+                    "time_events": _extract_events(span.events),
+                }
+            )
+            # TODO: Leverage more of the Cloud Trace API, e.g.
+            #  same_process_as_parent_span and child_span_count
+
+        return cloud_trace_spans
+
+    def shutdown(self):
+        pass
+
+
+def _get_hexadecimal_trace_id(trace_id: int) -> str:
+    return "{:032x}".format(trace_id)
+
+
+def _get_hexadecimal_span_id(span_id: int) -> str:
+    return "{:016x}".format(span_id)
+
+
+def _get_time_from_ns(nanoseconds: int) -> Dict:
+    """Given epoch nanoseconds, split into epoch milliseconds and remaining
+    nanoseconds"""
+    if not nanoseconds:
+        return None
+    seconds, nanos = divmod(nanoseconds, 1e9)
+    return {"seconds": int(seconds), "nanos": int(nanos)}
+
+
+def _get_truncatable_str_object(str_to_convert: str, max_length: int):
+    """Truncate the string if it exceeds the length limit and record the
+    truncated bytes count."""
+    truncated, truncated_byte_count = _truncate_str(str_to_convert, max_length)
+
+    return TruncatableString(
+        value=truncated, truncated_byte_count=truncated_byte_count
+    )
+
+
+def _truncate_str(str_to_check: str, limit: int) -> Tuple[str, int]:
+    """Check the length of a string. If exceeds limit, then truncate it."""
+    return str_to_check[:limit], max(0, len(str_to_check) - limit)
+
+
+def _extract_status(status: trace_api.Status) -> Status:
+    """Convert a Status object to protobuf object."""
+    if not status:
+        return None
+    status_dict = {"details": None, "code": status.canonical_code.value}
+
+    if status.description is not None:
+        status_dict["message"] = status.description
+
+    return Status(**status_dict)
+
+
+def _extract_links(links: Sequence[trace_api.Link]) -> ProtoSpan.Links:
+    """Convert span.links"""
+    if not links:
+        return None
+    extracted_links = []
+    for link in links:
+        trace_id = _get_hexadecimal_trace_id(link.context.trace_id)
+        span_id = _get_hexadecimal_span_id(link.context.span_id)
+        extracted_links.append(
+            {
+                "trace_id": trace_id,
+                "span_id": span_id,
+                "type": "TYPE_UNSPECIFIED",
+                "attributes": _extract_attributes(link.attributes),
+            }
+        )
+    return ProtoSpan.Links(link=extracted_links, dropped_links_count=0)
```

Comment:
Should the exporter enforce cloud trace's link and attribute count limits? `MAX_NUM_LINKS` is 32 -- lower than cloud trace's 128 -- and that's not likely to change, but since this is defined in the SDK and not the API it could theoretically cause problems for other SDKs with different limits.

Reply:
Hey Chris, a lot of limit stuff is addressed in branch stackdriver_event_limit, (including limits on #links, #events) which is rebased off of this branch. I had planned on merging them one after the other to avoid changing this PR anymore.

Reply:
This was fixed in another branch, stackdriver_event_limit, but I'll move the diffs into here